Framework/API/src/InfoComponentVisitor.cpp

```diff
@@ -0,0 +1,128 @@
+      m_detectorIdToIndexMapperFunction(mapperFunc) {
+  m_detectorIndices.reserve(nDetectors);
+}
+
+/**
+ * @brief InfoComponentVisitor::registerComponentAssembly
+ * @param assembly : ICompAssembly being visited
+ */
+void InfoComponentVisitor::registerComponentAssembly(
+    const ICompAssembly &assembly) {
+
+  std::vector<IComponent_const_sptr> assemblyChildren;
+  assembly.getChildren(assemblyChildren, false /*is recursive*/);
+
+  const size_t detectorStart = m_detectorIndices.size();
+  for (const auto &child : assemblyChildren) {
+    // register everything under this assembly
+    child->registerContents(*this);
+  }
+  const size_t detectorStop = m_detectorIndices.size();
+
+  m_ranges.emplace_back(std::make_pair(detectorStart, detectorStop));
+
+  // For any non-detector we extend the m_componetIds from the back
+  m_componentIds.emplace_back(assembly.getComponentID());
+}
+
+/**
+ * @brief InfoComponentVisitor::registerGenericComponent
+ * @param component : IComponent being visited
+ */
+void InfoComponentVisitor::registerGenericComponent(
+    const IComponent &component) {
+  /*
+   * For a generic leaf component we extend the component ids list, but
+   * the detector indexes entries will of course be empty
+   */
+  m_ranges.emplace_back(std::make_pair(0, 0)); // Represents an empty range
+  m_componentIds.emplace_back(component.getComponentID());
+}
+
+/**
+ * @brief InfoComponentVisitor::registerDetector
+ * @param detector : IDetector being visited
+ */
+void InfoComponentVisitor::registerDetector(const IDetector &detector) {
+
+  const auto detectorIndex =
+      m_detectorIdToIndexMapperFunction(detector.getID());
+
+  /* Unfortunately Mantid supports having detectors attached to an
+   * instrument that have an an invalid or duplicate detector id.
+   * We do not register detectors with the same id twice.
+   */
+  if (m_componentIds[detectorIndex] == nullptr) {
+
+    /* Already allocated we just need to index into the inital front-detector
+    * part of the collection.
+    * 1. Guarantee on grouping detectors by type such that the first n
+    * components
+    * are detectors.
+    * 2. Guarantee on ordering such that the
+    * detectorIndex == componentIndex for all detectors.
+    */
+    m_componentIds[detectorIndex] = detector.getComponentID();
+
+    // register the detector index
+    m_detectorIndices.push_back(detectorIndex);
+  }
+}
+
+/**
+ * @brief InfoComponentVisitor::componentDetectorRanges
+ * @return index ranges into the detectorIndices vector. Gives the
+ * intervals of detectors indices for non-detector components such as banks
+ */
+const std::vector<std::pair<size_t, size_t>> &
+InfoComponentVisitor::componentDetectorRanges() const {
+  return m_ranges;
+}
+
+/**
+ * @brief InfoComponentVisitor::detectorIndices
+ * @return detector indices in the order in which they have been visited
+ */
+const std::vector<size_t> &InfoComponentVisitor::detectorIndices() const {
```

Comment:
No. I've renamed this to `assemblySortedDetectorIndices()` in order for the range (start, stop) to work, we need to have an assembly grouped collection of detector indexes. We do not however need to use that when we are return detector indices (your point below).